Update query-string-syntax.asciidoc (#46863)
elastic/elasticsearch: docs/reference/query-dsl/query-string-syntax.asciidoc

If you need to use any of the characters which function as operators in your
query itself (and not as operators), then you should escape them with
a leading backslash. For instance, to search for `(1+1)=2`, you would

> need to write your query as `\(1\+1\)\=2`. In the case of using JSON format for the request body, two preceding backslashes (`\\`) required; since backslash is reserved escaping character in JSON strings.

Reviewer: ```suggestion
need to write your query as `\(1\+1\)\=2`. When using JSON for the request body, two preceding backslashes (`\\`) are required; the backslash is a reserved escaping character in JSON strings.
```

> The reserved characters are:  `+ - = && || > < ! ( ) { } [ ] ^ " ~ * ? : \ /`

Reviewer: Remove this sentence and relocate below the example.

Author: I've committed your changes. Which sentence should I remove?

Reviewer: Please remove the sentence at line 276. Thanks!

Author: And relocating it below the example? or completely removing it?

Reviewer: Completely remove the sentence at line 276

It looks like the same sentence was already added below the example with 28dffd427294aadfcf2c4ba36d4fc5eb6fd1e128. This would remove the duplicate.

Author: Oh! My bad.

[source,console]
----
GET /twitter/_search
Content-Type: application/json

{
  "query" : {
    "query_string" : {
      "query" : "kimchy\\!",
      "fields"  : ["user"]
    }
  }
}
----
// TEST[setup:twitter]

> Failing to escape these special characters correctly could lead to a syntax error which prevents your query from running. 

Reviewer: ```suggestion
The reserved characters are:  `+ - = && || > < ! ( ) { } [ ] ^ " ~ * ? : \ /`

Failing to escape these special characters correctly could lead to a syntax error which prevents your query from running. 
```

Failing to escape these special characters correctly could lead to a syntax
error which prevents your query from running.

NOTE: Since \ (backslash) is a special character in JSON strings, it needs 
to be escaped, hence two backslashes should be used.

> 

Reviewer 2: I think better than having two notes would be to make the former example even clearer. We could first extend the paragraph with the search for "(1+1)=2" by stating this would apply to keyword-field search mostly ("... to search for `(1+1)=2` in an unanalyzed `keyword` field ...") and then add another sentence that mentions the need for additional backslash escaping when using JSON (which is probably the most common way to use the REST API). Maybe even add an example query like:
```
GET /_search
{
  "query": {
    "query_string": {
      "query": "\\(1\\+1\\)\\=2"
    }
  }
}
```
wdyt?

Reviewer: I think a simpler example might illustrate the point better. Maybe something more like this:

```
====== Reserved characters

Query string syntax reserves the following characters as operators:

....
+ - = && || > < ! ( ) { } [ ] ^ " ~ * ? : \ /
....

To use one of these characters literally,
escape it with two preceding backlashes (`\\`).

For example,
the following search request uses a query string query
to match `kimchy!` in the `user` <<keyword,keyword>> field.
Two backslashes are required to escape `!` in `kimchy!`.

[source,console]
----
GET /twitter/_search
{
  "query" : {
    "query_string" : {
      "query" : "kimchy\\!",
      "fields"  : ["user"]
    }
  }
}
----
// TEST[setup:twitter]

NOTE: `>` and `<` can't be escaped at all. The only way to prevent them from
attempting to create a range query is to remove them from the query string
entirely.
```

@andrewnaguib Let me know if you think that addresses the issue. If so, please feel free to add it to your PR.

I think it might also be helpful to move this section higher in the file.

Author: Thanks @jrodewig and @cbuescher  for your response.  I believe both of you have provided a much more neat way on saying the same information. I will add them to the PR. I have a question tho, do you have internal policies for the documentation that I should be aware of, i.e., is it preferred, for instance, to adopt single example across the topic and so on? 

Moreover, @jrodewig you were suggesting moving the section in a higher position, I'm not sure where is the appropriate place except right after `Regular Expression` and before `Fuzziness` (I think its current place is better tho because all operators would have been mentioned already, i.e., a context of why they would be reserved was already shown. WDYT?

Reviewer: Thank you @andrewnaguib.

We don't have any specific policies about adopting the examples across topics. Generally, you should use an example you feel best fits the context. The `twitter` example provided is kind of an unofficial default, but please feel free to use your own if you feel it is a better fit.

I don't feel strongly about relocating the section. Your point about this section following others that mention operators is a good one. Let's just leave the section where it is for now.

Thanks again for your work on this.

Author: Thank you for your suggestions; I've added them.
